transaction/src/model/v1/attachments.rs

```diff
@@ -1,7 +1,177 @@
+
+#[derive(Debug, Clone, PartialEq, Eq, ManifestSbor)]
+pub struct PlaintextMessageV1 {
+    pub mime_type: String,
+    pub message: MessageContentsV1,
+}
+
+/// We explicitly mark content as either String or Bytes - this distinguishes (along with the mime type)
+/// whether the message is intended to be displayable as text, or not.
+///
+/// This data model ensures that messages intended to be displayable as text are valid unicode strings.
+#[derive(Debug, Clone, PartialEq, Eq, ManifestSbor)]
+pub enum MessageContentsV1 {
+    String(String),
+    Bytes(Vec<u8>),
+}
+
+impl MessageContentsV1 {
+    pub fn len(&self) -> usize {
+        match self {
+            MessageContentsV1::String(message) => message.len(),
+            MessageContentsV1::Bytes(message) => message.len(),
+        }
+    }
+}
+
+//============================================================================
+// ENCRYPTED MESSAGE
+//============================================================================
+
+/// A `PlaintextMessageV1` encrypted with "MultiPartyECIES" for a number of decryptors (public keys).
+///
+/// First, a `PlaintextMessageV1` should be created, and encoded as `manifest_sbor_encode(plaintext_message)`
+/// to get the plaintext message payload bytes.
+///
+/// The plaintext message payload bytes are encrypted via (128-bit) AES-GCM with an ephemeral symmetric key.
+///
+/// The (128-bit) AES-GCM symmetric key is encrypted separately for each decryptor public key via (256-bit) AES-KeyWrap.
+/// AES-KeyWrap uses a key derived via a KDF (Key Derivation Function) using a shared secret.
+/// For each decryptor public key, we create a shared curve point `G` via static Diffie-Helman between the
+/// decryptor public key, and a per-transaction ephemeral public key for that curve type.
+/// We then use that shared secret with a key derivation function to create the (256-bit) KEK (Key Encrypting Key):
+/// `KEK = HKDF(hash: Blake2b, secret: x co-ord of G, salt: [], length: 256 bits)`.
+///
+/// Note:
+/// - For ECDH, the secret we use is the `x` coordinate of the shared public point, unhashed. This ECDH output is
+///   known as ASN1 X9.63 variant of ECDH. Be careful - libsecp256k1 uses another non-standard variant.
+/// - We persist 128-bit symmetric keys because we wish to save on payload size, and:
+///   * 128-bit AES is considered secure enough for most use cases (EG bitcoin hash rate is only 2^93 / year)
+///   * It's being used with a transient key - so a hypothetical successful attack would only decrypt one message
+#[derive(Debug, Clone, PartialEq, Eq, ManifestSbor)]
+pub struct EncryptedMessageV1 {
+    pub encrypted: AesGcmPayload,
+    // Note we use a collection here rather than a struct to be forward-compatible to adding more curve types.
+    // The engine should validate each DecryptorsByCurve matches the CurveType.
+    pub decryptors_by_curve: BTreeMap<CurveType, DecryptorsByCurve>,
```

Comment:
How about?
```rust
decryptors: Vec<Decryptors>,

pub enum Decryptors {
    dh_ephemeral_public_key: PublicKey, // This is an enum
    decryptors: BTreeMap<PublicKeyFingerprint, AesWrapped128BitKey>,
}
```

`BTreeMap` breaks round-trip property.

Reply:
Will change to `IndexMap` to get the round-trip property back.